From: "Liam R. Howlett" <Liam.Howlett@oracle.com>
Subject: mm/mmap: fix __vma_adjust() memory leak

When shifting the arg pages, the maple state may be left with allocated
memory.  Free the memory by calling mas_destroy() unconditionally at the
end of the function.

Link: https://lkml.kernel.org/r/20220427224134.yurrplclbhmnrloj@revolver
Fixes: bd6a1fd58daf (mm: start tracking VMAs with maple tree)
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Reported-by: Qian Cai <quic_qiancai@quicinc.com>
Cc: Stephen Rothwell <sfr@canb.auug.org.au>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/mmap.c |    1 +
 1 file changed, 1 insertion(+)

--- a/mm/mmap.c~mm-start-tracking-vmas-with-maple-tree-fix
+++ a/mm/mmap.c
@@ -864,6 +864,7 @@ int __vma_adjust(struct vm_area_struct *
 	MA_STATE(mas, &mm->mm_mt, 0, 0);
 	struct vm_area_struct *exporter = NULL, *importer = NULL;
 
+	mas_destroy(&mas);
 	validate_mm(mm);
 	validate_mm_mt(mm);
 
_
